### DISCUSSION TITLE
Using a "walrus operator" for serial.read() error

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#15468] [Using MicroPython] [answered]
On an existing serial read of data from a battery management
system.
```py
while True:
    data = ser.read()

    print (f'{data}')

    if (data != None):  # need to do this test first
        var_sbms = data.decode('utf-8')

    if len(var_sbms) == 61:
        break  # good chance you got valid data
```

I get the following:
```
None
b';*3-(Y#hGRGWGXGUGTGTGRGU*?*N+#iz$@`#####################%N( \n'
```
I was told that the serial data comes from this system once a second.
Maybe the `None` indicates that I am not handling the data correctly.
However, doing the last test gives me the data I want.

Recently I was introduced to the "walrus operator", changing
the first two lines to:
```py
while (data := ser.read()):
    
    print (f'{data}')

    if (data != None):  # need to do this test first
        var_sbms = data.decode('utf-8')

    if len(var_sbms) == 61:
        break  # good chance you got valid data
```
Now gives me:
```
Traceback (most recent call last):
  File "main.py", line 47, in <module>
  File "sbms_logger.py", line 95, in <module>
  File "get_data.py", line 61, in get_data_from_SBMS0
  File "get_data.py", line 13, in dcmp
IndexError: string index out of range
```
I observe that the print statement never executes BUT the processing
seems to continue until the IndexError.

It appears that the walrus operator "spits-the-dummy" and I can not
catch the error with a try/except block.

What am I doing wrong?

**Comment by ricksorensen** [comment, marked as answer]
Presuming the error is because you fall through the loop and `data`  has no data, maybe what you want is:
```py
# read serial data
#   if data present, convert and check length
#       if length is expected, stop
#   keep trying
while True:
    if (data:=ser.read()):
        var_sbms = data.decode("utf-8")
        if len(var_sbms) == 61:
            break
```

In your original walrus form, the loop will break on the first read where ser.read() returns None/empty string- returning an empty `data` string.  Check the file and line-numbers to see when the string reference is failing.

**Reply by davefes** [reply]
Your suggestion does work.  Based on this discussion:
https://github.com/orgs/micropython/discussions/15397
and at the end I thought I should be able to do:
```py
while (data := ser.read()):
```
Thanks for pointing-out my wrong assumption.  What I didn't realise is I thought on `no data` I would see the break being executed after the print statement.
